app/src/ingest_bem_pdfs.py

```diff
@@ -76,15 +80,59 @@ def _parse_pdf(file: BinaryIO) -> list[EnrichedText]:
     for text in grouped_texts:
         text.id = str(uuid.uuid1())
     assert len(set(text.id for text in grouped_texts)) == len(grouped_texts)
+
     return grouped_texts
 
 
-def enrich_texts(file: BinaryIO) -> list[EnrichedText]:
+def enrich_texts(file: BinaryIO, unstructured_json: list[Element]) -> list[EnrichedText]:
+    enrich_text_list = []
     "Placeholder function. Will be implemented for DST-414, probably in a different file."
     outline: list[Heading] = pdf_utils.extract_outline(file)
-    for _heading in outline:
-        pass
-    return []
+    current_header = []
+    current_header_level = 1
+    for element in unstructured_json:
+        if element.category == "Footer" or element.category == "Header":
+            continue
+        if element.category == "Title":
+            header = match_heading(outline, element.text, element.metadata.page_number)
+            if header:
+                if header.level == 1:
+                    current_header = [header]
+                    current_header_level = 1
+                else:
+                    if header.title != current_header[-1].title:
+                        if current_header_level == header.level:
+                            current_header = current_header[:-1]
+                        if header.level > current_header_level:
+                            current_header_level = header.level
+                        current_header.append(header)
+
+        # Unstructured fails to categorize the date strings in the header,
+        # so manually check for that and ignore those too
+        if element.category == "UncategorizedText" and re.match(
+            r"^\d{1,2}-\d{1,2}-\d{4}$", element.text
+        ):
+            continue
+
+        enriched_text_item = EnrichedText(
+            text=element.text,
+            type=element.category,
+            page_number=element.metadata.page_number,
+            headings=current_header,
+            id=element._element_id,
+        )
+
+        enrich_text_list.append(enriched_text_item)
+    return enrich_text_list
+
+
+def match_heading(
+    heading_list: list[Heading], heading_name: str, page_number: int | None
+) -> Heading | None:
+    for heading in heading_list:
```

Comment:
FYI: Instead of iterating through all headings, I perform a stricter heading match algorithm that expects the next heading -- see https://github.com/navapbc/labs-decision-support-tool/blob/886b654922b62a7ad64aa11a1b3296dfe528c0f7/app/src/ingestion/pdf_stylings.py#L139.

This could help identify any mismatch between the outline (`heading_list`) and unstructure's parsed `Title` elements. Also note that it is possible (though unlikely) that the same heading text may appear multiple times in the outline and on the same page.